ci_cd/utils/versions.py

```diff
@@ -171,23 +197,72 @@ def patch(self) -> int:
         return self._patch
 
     @property
-    def pre_release(self) -> None | str:
+    def pre_release(self) -> str | None:
         """The pre-release part of the version
 
         This is the part supplied after a minus (`-`), but before a plus (`+`).
         """
         return self._pre_release
 
     @property
-    def build(self) -> None | str:
+    def build(self) -> str | None:
         """The build metadata part of the version.
 
         This is the part supplied at the end of the version, after a plus (`+`).
         """
         return self._build
 
+    @property
```

Comment:
To simplify the `as_python_version` method for better maintainability, you can refactor it to reduce complexity and improve readability. Here's a proposed approach:

1. **Separate Concerns:** Break down the method into smaller, focused sub-methods. Each sub-method should handle a specific aspect of the version conversion process.

2. **Clear Naming:** Use descriptive names for variables and methods to make the code self-explanatory.

3. **Avoid Deep Nesting:** Reduce nested if-else structures to make the code easier to follow.

Here's a refactored version of your method:

```python
def as_python_version(self, shortened: bool = True) -> Version:
    """Return the Python version as defined by `packaging.version.Version`."""
    if not self.python_version:
        return Version(self.shortened())

    version_parts = [self._get_epoch_segment(), self._get_release_segment(shortened)]
    version_parts.extend(self._get_additional_segments())

    return Version("".join(filter(None, version_parts)))

def _get_epoch_segment(self) -> str:
    """Return the epoch segment of the version."""
    return f"{self.python_version.epoch}!" if self.python_version.epoch != 0 else ""

def _get_release_segment(self, shortened: bool) -> str:
    """Return the release segment of the version."""
    return self.shortened() if shortened else ".".join(
        str(_) for _ in (self.major, self.minor, self.patch)
    )

def _get_additional_segments(self) -> List[str]:
    """Return the pre-release, post-release, development, and local segments."""
    segments = []
    if self._match_release_with_version():
        segments.extend(self._get_pre_post_dev_local_segments())

    return segments

def _match_release_with_version(self) -> bool:
    """Check if the release matches the current version."""
    return (self.major, self.minor, self.patch)[:len(self.python_version.release)] == self.python_version.release

def _get_pre_post_dev_local_segments(self) -> List[str]:
    """Get pre-release, post-release, development, and local segments if they exist."""
    segments = []
    if self.python_version.pre is not None:
        segments.append("".join(str(_) for _ in self.python_version.pre))

    if self.python_version.post is not None:
        segments.append(f".post{self.python_version.post}")

    if self.python_version.dev is not None:
        segments.append(f".dev{self.python_version.dev}")

    if self.python_version.local is not None:
        segments.append(f"+{self.python_version.local}")

    return segments
```

In this refactored version:

- The method `as_python_version` is now streamlined and delegates specific tasks to helper methods.
- Each helper method (`_get_epoch_segment`, `_get_release_segment`, `_get_additional_segments`, `_match_release_with_version`, and `_get_pre_post_dev_local_segments`) handles a distinct part of the version conversion process.
- This approach enhances readability and makes the code easier to maintain and modify in the future.
- The use of list comprehensions and `filter` helps to avoid complex nesting and makes the code more Pythonic.

Reply:
Sure. And I'd agree if SemanticVersion was all about redoing `packaging.version.Version`, but it is not. Well, it sort of is, but not in this sense. Adding this many utility methods to a class that already has a myriad of methods will most likely result in more confusion.

That said, I do like the last point of using comprehensions and `filter`. I'll try to implement this.

I've already tried to be as verbose as I can in this method, while still keeping it within the method.